Simplenote/Classes/ResultsController.swift

```diff
@@ -0,0 +1,170 @@
+import Foundation
+import CoreData
+
+
+// MARK: - Aliases
+//
+typealias ResultsChangeType = NSFetchedResultsChangeType
+typealias ResultsSectionInfo = NSFetchedResultsSectionInfo
+
+
+// MARK: - ResultsController
+//
+class ResultsController<T: NSManagedObject> {
+
+    /// FetchedResultsController Delegate Wrapper.
+    ///
+    private let internalDelegate = FetchedResultsControllerDelegateWrapper()
+
+    /// Results Controller
+    ///
+    private let resultsController: NSFetchedResultsController<T>
+
+    /// ResultsController's Fetch Request
+    ///
+    private var fetchRequest: NSFetchRequest<T> {
+        resultsController.fetchRequest
+    }
+
+    /// Filtering Predicate to be applied to the Results.
+    ///
+    var predicate: NSPredicate? {
+        get {
+            fetchRequest.predicate
+        }
+        set {
+            fetchRequest.predicate = newValue
+        }
+    }
+
+    /// Results's Sort Descriptor.
+    ///
+    var sortDescriptors: [NSSortDescriptor]? {
+        get {
+            fetchRequest.sortDescriptors
+        }
+        set {
+            fetchRequest.sortDescriptors = newValue
+        }
+    }
+
+    /// Closure to be executed before the results are changed.
+    ///
+    var onWillChangeContent: (() -> Void)?
+
+    /// Closure to be executed after the results are changed.
+    ///
+    var onDidChangeContent: (() -> Void)?
+
+    /// Closure to be executed whenever an Object is updated.
+    ///
+    var onDidChangeObject: ((_ object: Any, _ indexPath: IndexPath?, _ type: ResultsChangeType, _ newIndexPath: IndexPath?) -> Void)?
+
+    /// Closure to be executed whenever an entire Section is updated.
+    ///
+    var onDidChangeSection: ((_ sectionInfo: ResultsSectionInfo, _ sectionIndex: Int, _ type: ResultsChangeType) -> Void)?
+
+
+    /// Designated Initializer
+    ///
+    ///  - viewContext: Main Thread's MOC
+    ///  - sectionNameKeyPath: String containing the Section's Key Path
+    ///  - predicate: Filtering NSPredicate
+    ///  - sortDescriptors: Array of NSSortDescriptors which should be applied to the fetched results
+    ///
+    init(viewContext: NSManagedObjectContext,
+         sectionNameKeyPath: String? = nil,
+         matching predicate: NSPredicate? = nil,
+         sortedBy sortDescriptors: [NSSortDescriptor]) {
+
+        assert(viewContext.concurrencyType == .mainQueueConcurrencyType)
+
+        resultsController = {
+            let request = NSFetchRequest<T>(entityName: T.entityName)
+            request.predicate = predicate
+            request.sortDescriptors = sortDescriptors
+
+            return NSFetchedResultsController<T>(fetchRequest: request,
+                                                 managedObjectContext: viewContext,
+                                                 sectionNameKeyPath: sectionNameKeyPath,
+                                                 cacheName: nil)
+        }()
```

Comment:
same question here about the closure - why are we wrapping this vs. just writing it inline?

```swift
…

assert( … )

let request = NSFetchRequest<T>(entityName: T.entityName)
request.predicate = predicate
request.sortDescriptors = sortDescriptors

resultsController = NSFetchedResultsController<T>(
    fetchRequest: request,
    managedObjectContext: viewContext,
    sectionNameKeyPath: sectionNameKeyPath,
    cacheName: nil
)

setupResultsController()
setupDelegateWrapper()
```

Reply:
Oh! I see!. It's just a personal matter of readability. But as far as I know, the inner workings are pretty much the same!

Reply:
> just a personal matter of readability

if the compiler is smart enough. my point wasn't the indentation (I did that mostly to make it easier to read on GitHub)

the closure will have different runtime characteristics, though insignificant, but also debug implications since it's creating a function on the spot

Reply:
YYyyeah that's right. At the moment in which I wrote that, I felt quite strongly how bad this snippet looks like:

```
        let request = NSFetchRequest<T>(entityName: T.entityName)
        request.predicate = predicate
        request.sortDescriptors = sortDescriptors

        resultsController = NSFetchedResultsController<T>(fetchRequest: request,
                                                          managedObjectContext: viewContext,
                                                          sectionNameKeyPath: sectionNameKeyPath,
                                                          cacheName: nil)
```

But again, it's only cosmetic: the scope in which `request` lives is way bigger than it should be (it's _only_ required by `resultsController`, hence, the anonymous function declaration)